packages/ui/src/components/viewport/Coordinates.tsx

```diff
@@ -0,0 +1,90 @@
+import {useEffect, useState} from 'react';
+import {InfoBox} from '../controls/InfoBox';
+import {useCallback} from 'preact/hooks';
+import {useDocumentEvent} from '../../hooks';
+import {ViewportState} from './ViewportContext';
+
+interface CoordinatesProps {
+  visibility?: boolean;
+  viewState: ViewportState;
+  viewSize: DOMRectReadOnly;
+}
+
+export function Coordinates({
+  visibility,
+  viewState,
+  viewSize,
+}: CoordinatesProps) {
+  const [mousePos, setMousePos] = useState({x: 0, y: 0});
+  const [hover, setHover] = useState(false);
+
+  useEffect(() => {
+    const handleMouseMove = (event: {x: number; y: number}) => {
+      // This selector probably needs to be improved to select the overlay more reliably (future proof).
+      // Possibly use the internal state instead of a selector.
+      const canvas = document.getElementsByTagName('canvas')[1];
+      const rect = canvas.getBoundingClientRect();
+      const xPos = Math.round(
+        ((event.x - rect.x) / rect.width - 0.5) * canvas.width,
+      );
+      const yPos = Math.round(
+        ((event.y - rect.y) / rect.height - 0.5) * canvas.height,
+      );
```

Comment:
The origin being located in the center of the scene is explicit to the `Scene2D` class.
Other scenes may implement a different coordinate system.

We should extend the `Inspectable` interface so that scenes can transform the mouse position however they want.
This would look something like this:
```ts
const scene = useCurrentScene();
if (isInspectable(scene)) {
  // calculate the absolute mouse position the same way as in EditorPreview.tsx:
  const absoluteCoords = ... 
  // use the new method on the Inspectable interface to transform it:
  const point = scene.transformMousePosition(absoluteCoords);
}
```

Reply:
I understand what you mean here. But inside the Scene2D class how would I get the rect.x, rect.y, rect.width, and rect.height to perform the calculation?
```
public transformMousePosition(x: number, y: number): number | null {
  //   ???
}
```

Reply:
Would I just do this.getSize().x and this.getView().width?

Reply:
So the quoted code really only calculates the edge of the overlay window for the preview for bounds detection. The actual point calculation is done with...
```
let canvasPosX = event.x - state.size.x;
let canvasPosY = event.y - state.size.y;

canvasPosX -= state.x + state.width / 2;
canvasPosY -= state.y + state.height / 2;
canvasPosX /= state.zoom;
canvasPosY /= state.zoom;

canvasPosX = Math.round(canvasPosX);
canvasPosY = Math.round(canvasPosY);
```
But how would I get the Viewport state from within the scene's transformMousePosition without passing it as a parameter to the function?

Reply:
@aarthificial Can you provide me a bit more guidance on this one? I don't really know how to retrieve the ViewPort state from within the 'core' part of motion canvas. And when I try to use a transformMousePosition method (after added it as a interface and adding the method to the 2D scene class) my static analysis complains because...
```
const scene = useCurrentScene();
```
returns a Scene of type 'unknown'.

Reply:
Did you use the `isInspectable` function like this?
```ts
if (isInspectable(scene)) {
  // `scene` in this scope should be of type `Inspectable`
}
// outside the scope `scene` is back to being `Scene<unknown>`
```

Reply:
> Did you use the `isInspectable` function like this?
> 
> ```ts
> if (isInspectable(scene)) {
>   // `scene` in this scope should be of type `Inspectable`
> }
> // outside the scope `scene` is back to being `Scene<unknown>`
> ```

Yes, I did. It was considered to be of type 'Scene<unknown> & Inspectable'

Reply:
@aarthificial So basically my issues is...

It does not return the Scene type after I check if it is inspectable. Just scene type unknown. And I am not quite sure how to cast this scene to a specific type. Possibly a modification to the useCurrentScene function is required to return a scene with a specific type? I just am not sure how. Because I do not want to explicitly cast using...
const point = (scene as Scene2D).transformMousePosition(absoluteCoords);
because it is not guaranteed to always be Scene2D.

I 'think' that I figured out how to pull the viewport state in the same way as before with...
const state = useContext(ViewportContext);
with my issue being just an import path error. But I will have to wait and see on that one.

Reply:
I spoke with Hunter. He helped me resolve the issue. I did not know that I had to run a lerna build for cross package changes for the static analysis to update properly.

Reply:
I added the `transformMousePosition` function to the Inspectable interface and implemented it inside the Scene2D class. You will have to let me know if this implementation is sufficient.